lib/core/src/Cardano/Wallet/Network/Light.hs

```diff
@@ -0,0 +1,237 @@
+    :: (forall a. m a -> n a)
+    -> LightSyncSource m block addr txs
+    -> LightSyncSource n block addr txs 
+hoistLightSyncSource f x = LightSyncSource
+    { stabilityWindow = stabilityWindow x
+    , getHeader = getHeader x
+    , getTip = f $ getTip x
+    , isConsensus = f . isConsensus x
+    , getBlockHeaderAtHeight = f . getBlockHeaderAtHeight x
+    , getBlockHeaderAt = f . getBlockHeaderAt x
+    , getNextBlocks = f . getNextBlocks x
+    , getAddressTxs = \a b c -> f $ getAddressTxs x a b c
+    }
+
+type LightBlocks m block addr txs =
+    Either (NonEmpty block) (BlockSummary m addr txs)
+
+-- | Retrieve the 'ChainPoint' with the highest 'Slot'.
+latest :: [ChainPoint] -> ChainPoint
+latest [] = ChainPointAtGenesis
+latest xs = maximumBy compareSlot xs
+
+-- | Retrieve the 'ChainPoint' with the second-highest 'Slot'.
+secondLatest :: [ChainPoint] -> ChainPoint
+secondLatest []  = ChainPointAtGenesis
+secondLatest [_] = ChainPointAtGenesis
+secondLatest xs  = head . tail $ sortBy (flip compareSlot) xs
+
+-- | Drive a 'ChainFollower' using a 'LightSyncSource'.
+-- Never returns.
+lightSync
+    :: (Monad m, MonadDelay m)
+    => Tracer m LightLayerLog
+    -> LightSyncSource m block addr txs
+    -> ChainFollower m ChainPoint BlockHeader (LightBlocks m block addr txs)
+    -> m Void
+lightSync tr light follower = do
+    pts <- readLocalTip follower
+    syncFrom $ latest pts
+  where
+    idle = threadDelay secondsPerSlot
+    syncFrom pt = do
+        adventure <- boldyGoWhereNextPoint light pt
+        syncFrom =<< case adventure of
+            Rollback -> do
+                prev <- secondLatest <$> readLocalTip follower
+                -- NOTE: Rolling back to a result of 'readLocalTip'
+                -- should always be possible,
+                -- but the code here does not need this assumption.
+                traceWith tr $ MsgLightRollBackward pt prev
+                rollBackward follower prev
+            Stable old new tip -> do
+                let summary = mkBlockSummary light old new
+                traceWith tr $
+                    MsgLightRollForward (chainPointFromBlockHeader old) new tip
+                rollForward follower (Right summary) tip
+                pure $ chainPointFromBlockHeader new
+            Unstable blocks new tip -> do
+                case blocks of
+                    []     -> idle
+                    (b:bs) -> do
+                        traceWith tr $ MsgLightRollForward pt new tip
+                        rollForward follower (Left $ b :| bs) tip
+                pure $ chainPointFromBlockHeader new
+
+data NextPointAdventure block
+    = Rollback
+    -- ^ We are forced to roll back.
+    | Stable BlockHeader BlockHeader BlockHeader
+    -- ^ We are still in the stable region.
+    -- @Stable old new tip@.
+    | Unstable [block] BlockHeader BlockHeader
+    -- ^ We are entering the unstable region.
+    -- @Unstable blocks new tip@.
+
+boldyGoWhereNextPoint
```

Comment:
the name of this function does not need comment ;-)

Reply:
Are you ok with keeping it, or do you want me to change it? 🖖

Reply:
I thin we are doing here so serious stuff, that it kinda enforces proper naming;-) What about `proceedToNextPoint` ?